server/remote_cache/cachetools/cachetools.go

```diff
@@ -118,6 +119,90 @@ func GetBlob(ctx context.Context, bsClient bspb.ByteStreamClient, r *digest.Reso
 	return err
 }
 
+// BlobResponse is a response to an individual blob in a BatchReadBlobs request.
+type BlobResponse struct {
+	Digest *repb.Digest
+	Data   []byte
+
+	Err error
+}
+
+// BatchReadBlobs issues a BatchReadBlobs request and returns a mapping from
+// digest hash to byte payload.
+//
+// It validates the response so that if the returned err is nil, then all
+// digests in the request are guaranteed to have a corresponding map entry.
+func BatchReadBlobs(ctx context.Context, casClient repb.ContentAddressableStorageClient, req *repb.BatchReadBlobsRequest) (map[string]*BlobResponse, error) {
+	res, err := casClient.BatchReadBlobs(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	expected := map[string]struct{}{}
+	for _, d := range req.GetDigests() {
+		expected[d.GetHash()] = struct{}{}
+	}
+	// Validate that the response doesn't contain any unexpected digests.
+	for _, res := range res.Responses {
+		if _, ok := expected[res.GetDigest().GetHash()]; !ok {
+			return nil, status.UnknownErrorf("unexpected digest in batch response: %q", digest.String(res.GetDigest()))
+		}
+	}
+	// Build the results map, decompressing if needed and validating digests.
+	results := make(map[string]*BlobResponse, len(res.Responses))
+	for _, res := range res.Responses {
+		delete(expected, res.GetDigest().GetHash())
+
+		err := gstatus.ErrorProto(res.GetStatus())
+		if err != nil {
+			results[res.GetDigest().GetHash()] = &BlobResponse{Err: err}
+			continue
+		}
+		data := res.Data
+		// TODO: parallel decompression
+		// TODO: accept decompression buffer map as optional arg
+		if res.GetCompressor() == repb.Compressor_ZSTD {
+			buf := make([]byte, 0, res.GetDigest().GetSizeBytes())
+			d, err := compression.DecompressZstd(buf, res.Data)
+			if err != nil {
+				return nil, status.WrapError(err, "decompress blob")
+			}
+			data = d
+		}
+		// Validate digest
+		actualDigest, err := digest.Compute(bytes.NewReader(data), req.GetDigestFunction())
```

Comment:
nit: maybe call this downloadedContentDigest or dataDigest or something similar?

Reply:
Done